fq_test.go

```diff
@@ -0,0 +1,210 @@
+package fq
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+)
+
+// adapted from https://github.com/tadglines/wfq/blob/master/wfq_test.go
+
+type flowDesc struct {
+	// In
+	ftotal uint64 // Total units in flow
+	imin   uint64 // Min Packet size
+	imax   uint64 // Max Packet size
+
+	// Out
+	idealPercent  float64
+	actualPercent float64
+}
+
+func genFlow(fq *fqscheduler, desc *flowDesc, key uint64, done_wg *sync.WaitGroup) {
+	for i, t := uint64(1), uint64(0); t < desc.ftotal; i++ {
+		//time.Sleep(time.Microsecond)
+		it := new(Packet)
+		it.key = key
+		if desc.imin == desc.imax {
+			it.size = desc.imax
+		} else {
+			it.size = desc.imin + uint64(rand.Int63n(int64(desc.imax-desc.imin)))
+		}
+		if t+it.size > desc.ftotal {
+			it.size = desc.ftotal - t
+		}
+		t += it.size
+		it.seq = i
+		// new packet
+		fq.enqueue(it)
+	}
+	(*done_wg).Done()
+}
+
+func consumeQueue(t *testing.T, fq *fqscheduler, descs []flowDesc) (float64, error) {
+	active := make(map[uint64]bool)
+	var total uint64
+	acnt := make(map[uint64]uint64)
+	cnt := make(map[uint64]uint64)
+	seqs := make(map[uint64]uint64)
+
+	var wsum uint64
+	for range descs {
+		wsum += uint64(0 + 1)
+	}
+
+	time.Sleep(1 * time.Second)
+	for i, ok := fq.dequeue(); ok; i, ok = fq.dequeue() {
+		time.Sleep(time.Microsecond) // Simulate constrained bandwidth
+		it := i
+		seq := seqs[it.key]
+		if seq+1 != it.seq {
+			return 0, fmt.Errorf("Packet for flow %d came out of queue out-of-order: expected %d, got %d", it.key, seq+1, it.seq)
+		}
+		seqs[it.key] = it.seq
+
+		if cnt[it.key] == 0 {
+			active[it.key] = true
+		}
+		cnt[it.key] += it.size
+
+		if len(active) == len(descs) {
+			acnt[it.key] += it.size
+			total += it.size
+		}
+
+		if cnt[it.key] == descs[it.key].ftotal {
+			delete(active, it.key)
+		}
+	}
+
+	var variance float64
+	for key := uint64(0); key < uint64(len(descs)); key++ {
+		if total == 0 {
+			t.Fatalf("expected 'total' to be nonzero")
+		}
+		descs[key].idealPercent = (((float64(total) * float64(0+1)) / float64(wsum)) / float64(total)) * 100
+		descs[key].actualPercent = (float64(acnt[key]) / float64(total)) * 100
+
+		x := descs[key].idealPercent - descs[key].actualPercent
+		x *= x
+		variance += x
+	}
+
+	stdDev := math.Sqrt(variance)
+	return stdDev, nil
+}
+
+func TestSingleFlow(t *testing.T) {
+	runtime.GOMAXPROCS(runtime.NumCPU())
+	queues := initQueues(1, 1)
+	fq := newfqscheduler(queues)
+
+	go func() {
```

Comment:
It looks harmful to do this work in a separate goroutine rather than synchronously at this point.  The work here does not block, there is no reason not to do it synchronously.  Doing it asynchronously makes a correct test more difficult because you are not guaranteed when the enqueuing will finish.